### ISSUE TITLE
Memory Leaks

### ISSUE THREAD

**Issue by rabbiveesh** [Perl5-FFI/FFI-Platypus#190]
I have the following code (continuing my libpostal saga):
```perl
$ffi->attach(
  [ libpostal_get_default_options => 'expansion_defaults' ],
  [], 'expansion_options', );

$ffi->attach(
  [ libpostal_expansion_array_destroy => 'destroy_expansions' ],
  [ opaque => 'size_t' ]);

sub process_char_pp {
  my ($inner, @args) = @_;
  #call the function
  my $ret_ptr = 
    $inner->(@args, \my $ret_len);
  #nothing to do if there are no results
  return () unless $ret_len;
  #cast the char** into an arrayref
  my $ret = $ffi->cast( opaque => "string[$ret_len]", $ret_ptr);
  #clean up the C side of things
  destroy_expansions($ret_ptr, $ret_len);
  #send back the string[]
  return @$ret;
}

sub process_expansions {
  #third arg is the options, if not passed, then supply the
  #defaults
  $_[2] //= expansion_defaults();
  process_char_pp(@_);
}

$ffi->attach(
  [ libpostal_expand_address => 'expand_address' ],
  [ 'string', 'expansion_options', 'size_t*' ], 
  'opaque', \&process_expansions );
```

The problem I'm having is that none of the things returned from Platypus is ever getting freed. I did a simple test with `Test::LeakTrace` and it shows that the options struct leaks, as does the arrayref containing the returned strings and the strings themselves. (I also did a overview test, where I ran the function a million times, and continually ate up more and more memory).

Am I missing something? Where is the extra reference to each of these items?

The full code and the test file 't/03-valgrind.t' are in my repo at https://github.com/rabbiveesh/FFI-Postal

**Comment by plicease** [comment]
I believe this is fixed in #191 .  I plan on including an exhaustive set of regression tests in that PR.

**Comment by rabbiveesh** [comment]
It's getting there. I'm down to 3 leaks instead of the original 15. Thank you so much for the incredible work you put into this module!

**Comment by plicease** [comment]
Great.  There is another ticket for those leaks, which are less serious, so I'm closing this.  This is released as `0.97_05`:
https://metacpan.org/release/PLICEASE/FFI-Platypus-0.97_05
